specs/PushNotifications/PushNotifications-spec.md

```diff
@@ -0,0 +1,438 @@
+
+A Portal Registration flow through [AAD (Azure Active Directory)](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app) will also be defined that decouples the dependence of the Push Flow with the Partner Centre Portal. 
+The following are the main steps in the Push flow with AAD:
+* The developer will first obtain a Remote Identifer GUID and a client secret as part of this Portal Registration flow. This is a one time operation by the developer to provision an app. 
+* When requesting a channel using client Reunion APIs, the developer will pass in a remoteId and the WNS service will return a specially encoded Channel URL with that Remote Identifier. 
+* When the application service Pushes a notification down to the client, it will retrieve an access token with AAD using the client secret provisioned above. 
+* The application service will then pass that access token along with the payload in a Http POST operation to the channel URL. 
+* WNS will ensure that the payload is ultimately delivered to the app residing on the client device.
+
+Phase 1 (Reunion Preview):
+Our Push APIs will only support MSIX apps.
+
+Phase 2 (Reunion V1):
+Our Push APIs will add support for Unpackaged Win32.
+
+# Examples
+
+## Registration of the Push Activator for Win32 (Inproc)
+If a Win32 app needs to subscribe to Push triggers inproc, the code in Main would follow the pattern below:
+* The Registration will take in a CLSID of the COM server as part of the Activator setup. The registration API will simply be a thin wrapper around the BackgroundInfra WinRT APIs and abstract away the COM details from the developer.
+* The app will next query the ActivationEventArgs to check if the ActivationKind is of type Push
+* If the Activation is Push, the app will retrieve the IActivatedEventArgs and retrieve the Push payload from it. Care needs to be taken to Take a Deferral and Release the Deferral while processing the payload for Power Management compliance.
+* If the Activation is Foreground, the app will do 2 things:
+* * It will subscribe to a Foreground In-memory Push event
+* * It will request a Push Channel Asynchronously with an implementation of Progress and Completed event handler. 
+
+```cpp
+int main()
+{
+    // Always Register the Push trigger as the first step: It sets up the Inproc COM server for subsequent Push operations
+    std::wstring activatorGuid = L"{B8E58D9D-A229-44A8-B923-2823C3791BCE}";
+    InProcActivatorDetails details(activatorGuid);
+    PushNotificationChannelManager::RegisterPushNotificationActivator(details);
+
+    // Check to see if the WinMain activation is due to a Push Activator
+    auto args = AppLifecycle2::GetActivatedEventArgs();
+    auto kind = args.Kind2();
+
+    if (kind == ActivationKindExtension::Push)
+    {
+        auto activatedEventArgs = args.ActivatedArgs();
+        auto pushArgs = args.as<IPushNotificationActivatedEventArgs>();
```

Comment:
Nit: PushNotificationActivatedEventArgs

Reply:
You can cast it to the interface with c++ winRT : )